packages/cli/api-importers/openapi/openapi-ir-parser/src/asyncapi/AbstractAsyncAPIParserContext.ts

```diff
@@ -0,0 +1,170 @@
+import { OpenAPIV3 } from "openapi-types";
+
+import { Logger } from "@fern-api/logger";
+import { Namespace, SdkGroup, SdkGroupName } from "@fern-api/openapi-ir";
+import { TaskContext } from "@fern-api/task-context";
+
+import { ParseOpenAPIOptions } from "../options";
+import { SchemaParserContext } from "../schema/SchemaParserContext";
+import { SCHEMA_REFERENCE_PREFIX } from "../schema/convertSchemas";
+import { isReferenceObject } from "../schema/utils/isReferenceObject";
+
+export abstract class AbstractAsyncAPIParserContext<TDocument extends object> implements SchemaParserContext {
+    public readonly document: TDocument;
+    public readonly taskContext: TaskContext;
+    public readonly logger: Logger;
+    public readonly DUMMY: SchemaParserContext;
+    public readonly options: ParseOpenAPIOptions;
+    public readonly namespace: string | undefined;
+
+    protected static readonly MESSAGE_REFERENCE_PREFIX = "#/components/messages/";
+
+    constructor({
+        document,
+        taskContext,
+        options,
+        namespace
+    }: {
+        document: TDocument;
+        taskContext: TaskContext;
+        options: ParseOpenAPIOptions;
+        namespace: string | undefined;
+    }) {
+        this.document = document;
+        this.taskContext = taskContext;
+        this.logger = taskContext.logger;
+        this.DUMMY = this; // used by the schema logic
+        this.options = options;
+        this.namespace = namespace;
+    }
+
+    /**
+     * Merges the current namespace into the provided group name.
+     */
+    public resolveGroupName(groupName: SdkGroupName): SdkGroupName {
+        if (this.namespace != null) {
+            return [{ type: "namespace", name: this.namespace }, ...groupName];
+        }
+        return groupName;
+    }
+
+    /**
+     * Produces the final tag list (namespace + operation tags).
+     */
+    public resolveTags(operationTags: string[] | undefined, fallback?: string): SdkGroup[] {
+        if (this.namespace == null && operationTags == null && fallback != null) {
+            return [fallback];
+        }
+
+        const tags: SdkGroup[] = [];
+        if (this.namespace != null) {
+            const namespaceSegment: Namespace = {
+                type: "namespace",
+                name: this.namespace
+            };
+            tags.push(namespaceSegment);
+        }
+        return tags.concat(operationTags ?? []);
+    }
+
+    /**
+     * Resolves a schema reference of the form #/components/schemas/...
+     * Shared logic across v2 and v3 specs, so long as the shape matches:
+     *   this.document.components.schemas[...]
+     */
+    public resolveSchemaReference(schema: OpenAPIV3.ReferenceObject): OpenAPIV3.SchemaObject {
+        if (!schema.$ref.startsWith(SCHEMA_REFERENCE_PREFIX)) {
+            throw new Error(`Failed to resolve schema reference: ${schema.$ref}`);
```

Comment:
I think a better UX here is going to be to log an error and keep going? That way someone can still somewhat generate docs/SDKs even if their spec is slightly messed up

Reply:
@eyw520 might be doing a scan and reducing the number of errors thrown -- no issues if this is necessary though